### PR TITLE
Use next available key on duplicate coinbase commitment

### PR DESCRIPTION
**Pull request by bmhjyu**
If a coinbase commitment hits a duplication and there is no transactions to mine, it is possible that coinbase cannot be built for quite a long time. This change tries to build coinbase with an empty key identifier immediately in this case. The listening wallet will then use the next available key to build a coinbase commitment, which is different from the previous one and is unlikely to hit the duplication.

### CODE REVIEW

**Comment by garyyu** [comment]
Thanks for your 1st PR 👍 
I'm afraid I don't understand here `it is possible that coinbase cannot be built for quite a long time.`, do you see the real case?  and why `quite a long time`? I see there just a 100ms sleep there.

Could you please share more info?

**Comment by garyyu** [comment]
perhaps https://github.com/mimblewimble/grin/issues/1491 related.

File: servers/src/mining/mine_block.rs [hunk 1]
```diff
@@ -51,12 +51,15 @@ pub fn get_block(
...
 	while let Err(e) = result {
+		let mut new_key_id = key_id.to_owned();
 		match e {
 			self::Error::Chain(c) => match c.kind() {
 				chain::ErrorKind::DuplicateCommitment(_) => {
 					debug!(
 						"Duplicate commit for potential coinbase detected. Trying next derivation."
 					);
+					// use the next available key to generate a different coinbase commitment
+					new_key_id = None;
```
Comment (garyyu) [inline servers/src/mining/mine_block.rs]: 👍

**Comment by bmhjyu** [comment]
> Thanks for your 1st PR 
> I'm afraid I don't understand here `it is possible that coinbase cannot be built for quite a long time.`, do you see the real case? and why `quite a long time`? I see there just a 100ms sleep there.
> 
> Could you please share more info?

We have met this in production some time (stratum server stuck for hours). The commitment seems only depend on the amount and the key itself. If we hit the duplication and there is no transactions, the amount will always be just the coinbase reward itself. If we do not change the key, the commitment does not change until there is a transaction (amount changes as there are transaction fees).

**Pull request merged**